devito/types/grid.py

```diff
@@ -781,27 +784,27 @@ def __subdomain_finalize__(self, grid, counter=0, **kwargs):
         # compiler can use to generate code
         n = counter - npresets
         assert n >= 0
-        self._implicit_dimension = i_dim = Dimension(name='n%d' % n)
-        functions = []
-        for j in range(len(self._local_bounds)):
-            index = floor(j/2)
-            d = self.dimensions[index]
-            if j % 2 == 0:
-                fname = "%s_%s" % (self.name, d.min_name)
-            else:
-                fname = "%s_%s" % (self.name, d.max_name)
-            f = Function(name=fname, grid=self._grid, shape=(self._n_domains,),
-                         dimensions=(i_dim,), dtype=np.int32)
 
-            # Check if shorthand notation has been provided:
-            if isinstance(self._local_bounds[j], int):
-                f.data[:] = np.full((self._n_domains,), self._local_bounds[j],
-                                    dtype=np.int32)
-            else:
-                f.data[:] = self._local_bounds[j]
+        i_dim = Dimension(name='n%d' % n)
+
+        for i, d in enumerate(self.dimensions):
+            fnames = ["%s_%s" % (self.name, minmax)
+                      for minmax in (d.min_name, d.max_name)]
+            f = tuple(Function(name=fname, grid=self._grid, shape=(self._n_domains,),
+                               dimensions=(i_dim,), dtype=np.int32) for fname in fnames)
 
-            functions.append(f)
-        self._functions = as_tuple(functions)
+            # Check if shorthand notation has been provided:
+            for j in range(2):
+                idx = 2*i + j
+                if isinstance(self._local_bounds[idx], int):
+                    f[j].data[:] = np.full((self._n_domains,),
+                                           self._local_bounds[idx],
+                                           dtype=np.int32)
+                else:
+                    f[j].data[:] = self._local_bounds[idx]
+
+            d._implicit_dimension = i_dim
+            d._functions = f
```

Comment:
modifying `d` isn't great should rebuild it or create a new one

Reply:
this is within the lazy constructor, so it's kinda acceptable

Reply:
Was able to shuffle things around, remove the modification, and reduce the number of lines